### DISCUSSION TITLE
Closing the File [.exe] ForceFully

### DISCUSSION THREAD

**Discussion by ritishbhardwaj** [dennisvang/tufup#152] [Q&A] [answered]
I am running the application when i try to update it, the terminal which pops up after downloading the new .exe showing me the following error:
```  2024/06/20 16:22:18 ERROR 32 (0x00000020) Copying File C:\Users\RITISH~1.BHA\AppData\Local\Temp\tufup\pyleap.exe
The process cannot access the file because it is being used by another process.
Waiting 2 seconds... Retrying... 
 ```

Is there any way to delete this opened instance of the application forcefully just by making changes in tufup framework itself [in my environment]. I have seen some batch script that is written overthere, ```WIN_BATCH_DELETE_SELF = '(goto) 2>nul & del "%~f0"'```
but i am not sure what is this.

**Comment by dennisvang** [comment, marked as answer]
Hi @ritishbhardwaj,  this is *not* an error. 

The application is *already* closing at this point, because `sys.exit()` has been called:

https://github.com/dennisvang/tufup/blob/0283ebb420e7aa3408757dd9d36cd45c455c6490/src/tufup/utils/platform_specific.py#L234

It's just that the batch script, which runs in a separate process, has to wait for the application process to finish closing. Sometimes this takes a bit longer, causing the copy to fail, so what you're seeing is a simple retry mechanism with a short timeout. 

Checking if the app's process has stopped, properly and reliably, can become quite complicated, so, instead, we just retry if the copy fails.

If you want to change this, you go all out by specifying a custom batch template, as described in #48. However, you're on your own there.
